src/jose/JWT.js

```diff
@@ -25,245 +50,363 @@ class JWT extends JSONDocument {
-   * @param {String} data
-   * @param {Object} [options]
+   * @param {Object|String} data
+   * @param {String} [data.serialized] - Existing serialized JWT
    *
-   * @returns {Promise<JWT>}
+   * @return {JWT}
    */
-  static verify (key, data, options={}) {
+  static from (data) {
     let ExtendedJWT = this
-    let token
 
-    try {
-      token = ExtendedJWT.decode(data)
-    } catch (err) {
-      return Promise.reject(err)
+    // Decode serialized token
+    if (typeof data === 'string' || data.serialized) {
+      return this.decode(data.serialized || data)
     }
 
-    let { signatures } = token
+    let { payload, signatures, serialization } = data
 
-    // Assign options to JWT as nonenumerable properties
-    Object.keys(options).forEach(field => {
-      Object.defineProperty(token, field, { value: options[field], enumerable: false })
-    })
+    if (!payload) {
+      throw new DataError('Invalid JWT')
+    }
 
-    // Map keys to signatures by index
-    if (Array.isArray(key) && signatures) {
-      token.signatures = signatures.map((descriptor, index) => {
-        // TODO more sophisticated key mapping using hints like `kid'
-        if (index < key.length) {
-          Object.defineProperty(descriptor, 'key', { value: key[index], enumerable: false })
-        }
+    // Include compelete signature descriptors only
+    if (signatures && Array.isArray(signatures)) {
+      signatures = signatures.filter(descriptor => !descriptor.cryptoKey || descriptor.signature)
+    } else {
+      signatures = []
+    }
 
-        return descriptor
-      })
+    // Normalize existing flat signature
+    if (!data.cryptoKey && data.signature) {
+      let { protected: protectedHeader, header: unprotectedHeader, signature } = data
+      let descriptor = {}
+
+      if (!protectedHeader && unprotectedHeader) {
+        descriptor.protected = unprotectedHeader
+      } else {
+        descriptor.protected = protectedHeader
+        descriptor.header = unprotectedHeader
+      }
+
+      descriptor.signature = signature
 
-    // Assign single key to all signatures as nonenumerable property
-    } else if (signatures) {
-      token.signatures = signatures.map(descriptor => {
-        Object.defineProperty(descriptor, 'key', { value: key, enumerable: false })
-        return descriptor
+      if (signatures && Array.isArray(signatures)) {
+        signatures.unshift(descriptor)
+      } else {
+        signatures = [descriptor]
+      }
+    }
+
+    return new ExtendedJWT(
+      clean({
+        payload,
+        signatures,
+        serialization,
+        type: 'JWS'
       })
+    )
+  }
 
-    // Assign key to token as nonenumerable property
-    } else {
-      Object.defineProperty(token, 'key', { value: key, enumerable: false })
+  /**
+   * sign
+   *
+   * @description
+   * Sign a JSON Web Token
+   *
+   * @params {...Object} data - Token data
+   *
+   * @returns {Promise<SerializedToken>}
+   */
+  static sign (...data) {
```

Comment:
The intended use of the ```...data``` parameter is that you can pass any combination of objects/serializations into the sign method and it will simply "deal with it" for you.

For example in examples/RS256-JWT-Sign.js:
```js
JWT.sign({ header, payload }, { serialization: 'compact', cryptoKey: privateKey })
JWT.sign({ protected: header, payload }, { serialization: 'flattened', cryptoKey: privateKey })
JWT.sign({ signatures: [{ protected: header, cryptoKey: privateKey }, { protected: header, cryptoKey: privateKey }], payload }, { serialization: 'json' })
JWT.sign({ serialized }, { signatures: [{ protected: header, cryptoKey: privateKey }] }, { serialization: 'json' })
```

All it means is that you can pass any number of Object arguments in and they all get shallow merged (with ```Object.assign```) into a single object before proceeding.

This should be useful if -- for example -- you already have some token data from somewhere and you wish to override the output serialization or provide the signing key. You might write something like this:
```js
JWT.sign(myTokenData, { serialization: 'document', cryptoKey: privateKey })
```

as opposed to:

```js
myTokenData.serialization = 'document'
myTokenData.cryptoKey = privateKey

JWT.sign(myTokenData)
```

It also future-proofs our API as any additional parameters can be added without breaking backwards compatibility; at least for most additional paramaters we _may_ want to add in the future.